SPIFlashLogger.class.nut

```diff
@@ -76,53 +114,191 @@ class SPIFlashLogger {
+    // each of which gives access to an integer value:
+    //   size    - The size of the SPIFlash in bytes
+    //   len     - The number of bytes allocated to the logger
+    //   start   - The first byte used by the logger
+    //   end     - The last byte used by the logger
+    //   sectors - The number of sectors allocated to the logger
+    //   sectorSize - The size of sectors in bytes
     function dimensions() {
         return { "size": _size, "len": _len, "start": _start, "end": _end, "sectors": _sectors, "sector_size": SPIFLASHLOGGER_SECTOR_SIZE }
     }
 
+    // Writes any serializable object to the memory allocated
+    // for the SPIFlashLogger. If the memory is full,
+    // the logger begins overwriting the oldest entries.
+    // If the provided object can not be serialized,
+    // the exception is thrown by the underlying serializer class.
+    //
+    // Parameters:
+    //    object - an object to serialize and save
     function write(object) {
         // Check of the object will fit
-        local obj_len = _serializer.sizeof(object, SPIFLASHLOGGER_OBJECT_MARKER);
-        if (obj_len > _max_data) throw "Cannot store objects larger than alloted memory."
+        local objLength = _serializer.sizeof(object, SPIFLASHLOGGER_OBJECT_MARKER);
+        if (objLength > _maxData) throw "Cannot store objects larger than alloted memory."
 
         // Serialize the object
         local obj = _serializer.serialize(object, SPIFLASHLOGGER_OBJECT_MARKER);
 
         _enable();
 
         // Write one sector at a time with the metadata attached
-        local obj_pos = 0;
-        local obj_remaining = obj_len;
+        local objPos = 0;
+        local objRemaining = objLength;
         do {
 
             // How far are we from the end of the sector
-            if (_at_pos < SPIFLASHLOGGER_SECTOR_META_SIZE) _at_pos = SPIFLASHLOGGER_SECTOR_META_SIZE;
-            local sec_remaining = SPIFLASHLOGGER_SECTOR_SIZE - _at_pos;
-            if (obj_remaining < sec_remaining) sec_remaining = obj_remaining;
+            if (_atPos < SPIFLASHLOGGER_SECTOR_METADATA_SIZE) _atPos = SPIFLASHLOGGER_SECTOR_METADATA_SIZE;
+            local secRemaining = SPIFLASHLOGGER_SECTOR_SIZE - _atPos;
+            if (objRemaining < secRemaining) secRemaining = objRemaining;
 
             // We are too close to the end of the sector, skip to the next sector
-            if (obj_pos == 0 && sec_remaining < SPIFLASHLOGGER_OBJECT_MIN_SIZE) {
-                _at_sec = (_at_sec + 1) % _sectors;
-                _at_pos = SPIFLASHLOGGER_SECTOR_META_SIZE;
+            if (objPos == 0 && secRemaining < SPIFLASHLOGGER_OBJECT_MIN_SIZE) {
+                _atSec = (_atSec + 1) % _sectors;
+                _atPos = SPIFLASHLOGGER_SECTOR_METADATA_SIZE;
+            }
+            // Handle overflow use-case for a one-sector logger
+            else if (_sectors == 1 && objPos == 0 && objRemaining > secRemaining) {
```

Comment:
If object does not fit to the remaining sectors space for a 1-secrot logger then we should not write it partially.  => we need to erase single sector and write down object into erased sector

Reply:
can objects be partially be written to multiple sectors?

Reply:
if so, how does it work? what if the remainder doesn't fit the available space in the second sector?

Reply:
1. If we are talking about current code the it works for 1 sector logger only

2. For a multiple sectors logger, it is possible to split object between multiple sectors. There is only one limitation for that:  start code + length (6 bytes) should fit into the first sector.
we have a special test case for that when start code is in a first sector and object payload is in a second sector.

3. We have an exception when object does not fit into the logger's sector. So, it is not possible to start writing object which does not fit into the total available sectors

4. And it seems that I"ve got your point, we have the same problem for a multiple sectors.
For example, if we have 2 sectors logger, and we are trying to write down 1,5 sectors size object, then we can get the same problem.